### PR TITLE
kep: update arena v2 proposal

### PR DESCRIPTION
**Pull request by vicoooo26**
## Purpose of this PR

Follow-up to #1441. Revises the Arena v2 KEP documents based on review feedback — sharpens the CLI flag mapping, simplifies the YAML schema, and tightens the proposal README so the motivation and scope are clearer before implementation starts.

**Proposed changes:**

- Refine `cli-flag-mapping.md`: reorganize flag categories and clarify the mapping between legacy v1 flags and their v2 equivalents.
- Simplify `yaml-schema.md`: drop redundant fields, consolidate schema sections, and align the spec with the current proposal direction.
- Revise `README.md`: clarify motivation, scope, and design rationale for Arena v2.

## Change Category

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- [x] Documentation update

### Rationale

The initial Arena v2 proposal (#1441) landed as a first draft. This PR incorporates follow-up revisions to the proposal artifacts before any implementation work begins, so reviewers and contributors have an up-to-date design reference.

Refs: #1441

### CODE REVIEW

File: keps/arena-v2/yaml-schema.md [hunk 1]
```diff
@@ -461,149 +472,18 @@ Init containers do not support `envs`, `secrets`, or `resources` configuration.
 
 ```
 User writes: worker.replicas: 4
-  pytorch    → PyTorchJob  { Master(1) + Worker(3) }   = 4 pods
+  pytorch    → PyTorchJob  { Master(1) + Worker(4) }   = 5 pods
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:537]: The replica semantics changed significantly here: `worker.replicas: 4` now means 5 total pods (Master + 4 Workers) instead of the previous 4 pods (Master + 3 Workers).

This is actually more intuitive and aligns with the PyTorchJob CRD API (Master and Worker are independent ReplicaSpecs). However, v1 users migrating to v2 might be tripped up by this difference.

Suggestion: consider adding a short migration note in the README's Upgrade Strategy section explaining this semantic change, e.g. "In v2, `worker.replicas` specifies the actual worker count; the master is always an additional pod."

File: keps/arena-v2/cli-flag-mapping.md [hunk 2]
```diff
@@ -1,31 +1,33 @@
-# Arena V1 → V2 Flag Audit
+# Arena V1 Flag → V2 YAML Schema Audit
 
 This document maps all arena v1 training CLI flags to arena v2 YAML schema coverage.
 
 **Scope:** `arena submit` training subcommands  
 **Source of truth:** arena v2 YAML Schema + arena v1 CLI
 
+> **Note:** v2 CLI flag-to-YAML override mechanism will be designed separately as a generic `--set`-style approach (similar to Helm). This document tracks v1 flag → v2 YAML schema coverage only.
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: Good to separate the CLI override design. Would it be helpful to link a tracking issue here so readers can follow the `--set`-style override design progress? That way this document stays self-contained as a schema coverage reference while still pointing people to the CLI override work.

File: keps/arena-v2/README.md [hunk 3]
```diff
@@ -191,36 +191,35 @@ The test plan for Arena v2 includes:
...
     - `serve`
-- Add DeepSpeed support and example
-- TensorFlow provider
-- Ray provider (RayJob CRD)
+- Add DeepSpeed, TensorFlow, Ray support and examples
 
 **Stable (Phase 3 - future):**
 
-- Platform agnostic configuration via `arena configure`
+- CLI flag override support
```
Comment (cheyang) [inline keps/arena-v2/README.md]: Moving CLI flag override to Stable phase means Alpha and Beta users are limited to pure YAML input. For quick iteration scenarios (e.g. just bumping GPU count), editing YAML every time adds friction.

Have you considered introducing at least a minimal `--set key=value` override in Beta? Or is there a strong design reason to defer it entirely to Stable (e.g. wanting the YAML schema to fully stabilize first)?

File: keps/arena-v2/yaml-schema.md [hunk 4]
```diff
@@ -411,29 +419,32 @@ arena job resume <name>    # Resume task
...
 
   - hdfs: hdfs://namenode:8020/models/resnet
+    local_path: /models
     mounts:
     - name: checkpoints
       mount_path: /models
 ```
 
-Each entry creates an initContainer + emptyDir volume. The emptyDir is mounted at `mount_path` in both the initContainer and the main container across all pods.
+Each entry creates an initContainer and volumes inherited from `storages`. The `mounts` field overrides these inherited configurations by name, and the final volumes are mounted at `mount_path` in both the initContainer and main container across all pods.
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:488]: The new model where sync entries inherit volumes from `storages` and `mounts` overrides by name is cleaner than the old per-entry emptyDir approach. However the override-by-name semantics could use a brief example showing what happens when a `mounts[].name` matches a `storages` entry vs. when it doesn't.

A 2-3 line example would make this immediately clear for implementers.

File: keps/arena-v2/cli-flag-mapping.md (hunk 2 above)
Comment (vicoooo26) [reply keps/arena-v2/cli-flag-mapping.md]: A tracking issue will be created after the v2 MVP implementation is merged.

File: keps/arena-v2/README.md (hunk 3 above)
Comment (vicoooo26) [reply keps/arena-v2/README.md]: We can support overriding of stable YAML fields (such as name, worker.replicas, etc.) during the beta phase. Full flag support will be added once the YAML schema is finalized.

File: keps/arena-v2/yaml-schema.md (hunk 1 above)
Comment (vicoooo26) [reply keps/arena-v2/yaml-schema.md:537]: Added additional clarification about the GPU-using pods in `PyTorch`.

File: keps/arena-v2/yaml-schema.md (hunk 4 above)
Comment (vicoooo26) [reply keps/arena-v2/yaml-schema.md:488]: Added additional clarification about `mounts` behavior in `sync` block.

**Comment by cheyang** [comment]
@gemini-code-assist /review

**Comment by cheyang** [comment]
Thanks for the follow-up revision — this draft is significantly tighter than #1441 and the CLI/schema split is much clearer. Most of the cross-document inconsistencies I noted in the first round are already resolved here.

Before this lands, I'd like to see the per-role config wording aligned: the README says per-role config uses `chief/ps/evaluator/launcher` blocks, while `yaml-schema.md` only documents a `master` block with a note that it 'may also be declared as' those other roles. Clarifying the actual YAML keys (or unifying on `master`) will prevent implementers from guessing. There are also a few small typos and one phase-scope mismatch (TFJob provider tested in Alpha but scheduled for Beta) to tidy up. I've left inline comments for these; once they're addressed this should be good to go.

File: keps/arena-v2/README.md [hunk 5]
```diff
@@ -154,11 +156,25 @@ PyTorchJob CRD
 | **Helm charts** | 24 charts maintained alongside Go code | 0 |
 | **Temp files per submission** | 3 (values.yaml, rendered manifest, app-info) | 0 |
 | **Resource lifecycle** | 3-step delete (read ConfigMap → delete resources → delete ConfigMap) | 1-step delete (delete anchor resource → K8s GC cascades) |
-| **Per-role config (TFJob)** | 40+ CLI flags (e.g., `--worker-image`, `--ps-cpu`, `--chief-memory`) | `chief/ps` block in YAML with independent per-role declaration |
+| **Per-role config** | 40+ CLI flags (e.g., `--worker-image`, `--ps-cpu`, `--chief-memory`) | `chief/ps/evaluator/launcher` block in YAML with independent per-role declaration |
```
Comment (cheyang) [inline keps/arena-v2/README.md]: The README states per-role config uses `chief/ps/evaluator/launcher` blocks, but `yaml-schema.md` only defines a `master` block and says it 'may also be declared as launcher, chief, ps, or evaluator depending on the framework'. This is ambiguous: readers cannot tell whether the YAML key is always `master` or whether it changes per framework. Please either define the actual role keys in the schema (e.g. `chief`, `ps`, `evaluator`, `launcher`) or align the README to use `master` consistently and explain the framework-specific mapping explicitly.

File: keps/arena-v2/yaml-schema.md [hunk 6]
```diff
@@ -50,32 +49,43 @@ envs:                                # Optional. Common envs.
 worker:                             # Required
   replicas: 4                       # Required. Worker replicas must be greater than 0
   envs:                             # Optional
-    NCCL_DEBUG: DEBUG               # Worker: inherited env + NCCL_DEBUG (override/merge)
+    NCCL_DEBUG: DEBUG               # Merges with top-level envs (worker value overrides)
   resources:                        # Required
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md]: The schema marks `worker.resources` as Required and the comment says 'Worker: manual specification required'. For CPU-only training jobs, specifying a GPU resource is unnecessary. Please clarify whether `resources` is required as a block, which sub-fields are required, and how a CPU-only job should declare `worker.resources` (e.g. empty, or with `cpu`/`memory` only).

File: keps/arena-v2/README.md [hunk 7]
```diff
@@ -191,36 +207,36 @@ The test plan for Arena v2 includes:
...
+- Dry-run mode for CRD inspection
 
 **Beta (Phase 2 - next):**
 
-- Dry-run mode for CRD inspection
 - Advanced features: `lifecycle` policies, `sync`/`init` containers
 - Users begin migrating simple jobs to v2
 - Comprehensive CLI flag coverage matching v1 functionality
 - Output formats: JSON, YAML, wide
 - InferenceTask kind (RBG only)
```
Comment (cheyang) [inline keps/arena-v2/README.md]: The Beta section says 'InferenceTask kind (RBG only)'. What does 'RBG' stand for? If it is a typo or internal acronym, please expand or correct it so external readers understand the scope.

File: keps/arena-v2/README.md [hunk 8]
```diff
@@ -296,13 +312,13 @@ The test plan for Arena v2 includes:
 
 **How can an operator determine if the feature is working correctly?**
 
-- v2 provides `--dry-run` flag for inspecting generated CRDs before submission
+- v2 will provide `--dry-run` flag for inspecting generated CRDs before submission
```
Comment (cheyang) [inline keps/arena-v2/README.md]: The PRR questionnaire now uses future tense ('v2 will provide `--dry-run`'), but the Troubleshooting section later says '--dry-run flag shows generated CRD before submission' in present tense. Is the present-tense Troubleshooting entry intended to describe the eventual behavior, or should it also be updated to future tense until the feature lands?

File: keps/arena-v2/README.md [hunk 9]
```diff
@@ -191,36 +207,36 @@ The test plan for Arena v2 includes:
 - Legacy arena `submit` command for backward compatibility
 - Core training submission (PyTorchJob, MPIJob) with YAML schema
 - Basic command (`arena version`, `arena help`)
-- Display cluster resource info（`arena top`）
+- Display cluster resource info (`arena top`)
 - Basic job manipulates and status queries (`arena job delete`, `arena job get`, `arena job list`, `arena job status`, `arena job logs`), using `arena job` prefix alias to support v1 `arena delete/get/list/status/logs` commands
```
Comment (cheyang) [inline keps/arena-v2/README.md]: A few small typos reduce polish: line 211 'Basic job manipulates' should be 'Basic job manipulation'; line 214 uses a full-width Chinese comma '，' after `run`; line 138 'respects training job CR `ttlSecondsAfterFinished`' should be 'respects the training job CR's `ttlSecondsAfterFinished`'. A quick cleanup pass would help.

File: keps/arena-v2/README.md (hunk 8 above)
Comment (vicoooo26) [reply keps/arena-v2/README.md]: The `--dry-run` flag has been implemented as part of the MVP. I'll remove the future tense.

File: keps/arena-v2/README.md (hunk 9 above)
Comment (vicoooo26) [reply keps/arena-v2/README.md]: Fixed

File: keps/arena-v2/README.md (hunk 7 above)
Comment (vicoooo26) [reply keps/arena-v2/README.md]: Added a link to the RBG repo and explained its usage.

**Comment by cheyang** [comment]
@gemini-code-assist /review

File: keps/arena-v2/yaml-schema.md [hunk 10]
```diff
@@ -50,32 +49,45 @@ envs:                                # Optional. Common envs.
...
-    NCCL_DEBUG: DEBUG               # Worker: inherited env + NCCL_DEBUG (override/merge)
-  resources:                        # Required
-    nvidia.com/gpu: 8               # Worker: manual specification required
+    NCCL_DEBUG: DEBUG               # Merges with top-level envs (worker value overrides)
+  resources:                        # Recommended, but not required
+    nvidia.com/gpu: 8               # Worker: manual specification
+    cpu: 32
+    memory: 128Gi
+
+# The master field may also be declared as launcher, chief, ps, or evaluator depending on the framework; replicas, resources, and envs settings may vary accordingly.
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md]: The per-role YAML configuration is still underspecified. `README.md` and the flag-mapping tables talk about independent `chief/ps/evaluator/launcher` blocks, but the schema only documents a `worker` block and a generic `master` alias. Please add an explicit per-role schema section that lists the valid top-level role keys per framework, clarifies whether `master` is a literal key or a placeholder for `launcher/chief/ps/evaluator`, and shows how defaults and overrides flow from `worker` to each role. This is the remaining item from the previous review.

File: keps/arena-v2/README.md [hunk 11]
```diff
@@ -70,14 +72,14 @@ Arena v2 adopts a modular architecture that separates concerns and reduces coupl
...
 ```
 cmd/arena-v2/main.go          → Entry point
-pkg/cli/                       → Cobra commands (root, run, get, list, delete, stop, resume, logs, submit, version, completion)
+pkg/cli/                       → Cobra commands (root, status, top, check, job, logs, submit, version, completion)
 pkg/task/                      → Task data model, YAML loader, flag overrides
-pkg/provider/                  → Provider interface + implementations (PyTorchJob, TFJob, MPIJob)
+pkg/provider/                  → Provider interface + implementations (PyTorchJob, MPIJob)
 pkg/client/                    → K8s dynamic + typed client wrapper
 pkg/output/                    → Table formatting for CLI output
 ```
```
Comment (cheyang) [inline keps/arena-v2/README.md:80]: The architecture overview lists providers as `PyTorchJob, MPIJob`, but the Provider Interface sentence still includes `TFJob`. Since the TFJob provider is planned for Beta, please keep these two statements consistent so readers don't think TFJob is already implemented in Alpha.

File: keps/arena-v2/README.md [hunk 12]
```diff
@@ -302,7 +318,7 @@ The test plan for Arena v2 includes:
 
 **How can this feature be monitored?**
 
-- Kubernetes metrics for created CRDs (PyTorchJob, TFJob, MPIJob)
+- Kubernetes metrics for created CRDs (PyTorchJob, MPIJob)
```
Comment (cheyang) [inline keps/arena-v2/README.md:326]: The monitoring section lists `TFJob` metrics, but the Alpha phase only covers PyTorchJob and MPIJob. Consider qualifying this as applicable once the TFJob provider lands in Beta.

File: keps/arena-v2/yaml-schema.md [hunk 13]
```diff
@@ -50,32 +49,45 @@ envs:                                # Optional. Common envs.
...
+master:                             # Optional. Default: same as worker, except replicas is fixed to 1.
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md]: Line 58 says replicas/resources/envs may vary by framework, while the `master` block comment says `replicas` is fixed to 1. Clarify which role types allow configurable replicas and which are fixed at 1.

File: keps/arena-v2/yaml-schema.md (hunk 6 above)
Comment (vicoooo26) [reply keps/arena-v2/yaml-schema.md]: Updated.

**Comment by cheyang** [comment]
Thanks for the quick update. The new `framework-specific configuration` section in `yaml-schema.md` resolves the per-role YAML key alignment issue from the previous review, and the test plan and provider interface are now consistent with the Alpha/Beta phase definitions. With the typos and wording also cleaned up, this KEP looks coherent and complete. Approving — please consider the one inline nit about `worker.replicas` validation at your discretion.

File: keps/arena-v2/yaml-schema.md [hunk 14]
```diff
@@ -47,35 +46,49 @@ envs:                                # Optional. Common envs.
   NCCL_IB_DISABLE: "0"
 
 # ─── Scale ───
-worker:                             # Required
-  replicas: 4                       # Required. Worker replicas must be greater than 0
+worker:                             # Required for non-PyTorch frameworks
+  replicas: 4                       # Required if the worker block is present. Must be ≥ 0.
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md]: The schema comment says `worker.replicas` must be `≥ 0`, but the provider mapping only explains that `0` is invalid when the PyTorch `master` block is omitted. Consider clarifying whether `0` is ever valid for non-PyTorch frameworks such as MPI or TensorFlow.

File: keps/arena-v2/README.md (hunk 5 above)
Comment (vicoooo26) [reply keps/arena-v2/README.md]: Added a new framework-specific configuration section to `yaml-schema.md` that details framework roles and behaviors.

File: keps/arena-v2/yaml-schema.md (hunk 14 above)
Comment (vicoooo26) [reply keps/arena-v2/yaml-schema.md]: If the `worker` block is specified, `replicas` must be > 0. Since this field is optional for PyTorch, the user can simply omit it if the `worker` role isn't needed.

File: keps/arena-v2/README.md (hunk 11 above)
Comment (vicoooo26) [reply keps/arena-v2/README.md:80]: `TFJob` is currently excluded from the MVP phase.

File: keps/arena-v2/README.md (hunk 12 above)
Comment (vicoooo26) [reply keps/arena-v2/README.md:326]: TFJob is currently excluded from the MVP phase.

File: keps/arena-v2/yaml-schema.md (hunk 13 above)
Comment (vicoooo26) [reply keps/arena-v2/yaml-schema.md]: Add a new framework-specific configuration section for clarity.

**Comment by cheyang** [comment]
This revision addresses the cross-document consistency issues raised in the first review round. The framework-specific configuration section, the migration note for PyTorch `worker.replicas`, and the Alpha/Beta phase scoping are all much clearer now. I don't see any blockers. The remaining items are minor wording/validation clarifications that can be handled at the author's discretion.

File: keps/arena-v2/yaml-schema.md [hunk 15]
```diff
@@ -47,62 +46,85 @@ envs:                                # Optional. Common envs.
   NCCL_IB_DISABLE: "0"
 
 # ─── Scale ───
-worker:                             # Required
-  replicas: 4                       # Required. Worker replicas must be greater than 0
+worker:                             # Required for non-PyTorch frameworks
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md]: The schema marks `worker` as required for every non-PyTorch framework. For TensorFlow this could be overly restrictive once the provider lands: a chief-only TFJob is valid in the Training Operator but would still be rejected here because no `worker` block is present. Consider relaxing this (or at least noting the exception) when the TFJob provider is designed in Beta.

File: keps/arena-v2/README.md [hunk 16]
```diff
@@ -167,7 +183,7 @@ The test plan for Arena v2 includes:
```
Comment (cheyang) [inline keps/arena-v2/README.md:183]: The E2E test plan could explicitly cover the two new edge cases introduced by this revision: (1) PyTorch job with `worker` omitted and `master` specified, and (2) MPI job relying on the default CPU-only launcher configuration. These are the main behavioral boundaries added by the framework-specific sections.

File: keps/arena-v2/yaml-schema.md [hunk 17]
```diff
@@ -47,62 +46,85 @@ envs:                                # Optional. Common envs.
...
-  resources:                        # Required
-    nvidia.com/gpu: 8               # Worker: manual specification required
+    NCCL_DEBUG: DEBUG               # Merges with top-level envs (worker value overrides)
+  resources:                        # Recommended, but not required. Pod resources will be unset if this block is omitted.
+    nvidia.com/gpu: 8
+    cpu: 32
+    memory: 128Gi
+
+# Other fields, such as launcher, chief, ps, or evaluator, may be available depending on the framework. 
+# Settings for replicas, resources, and environment variables may vary accordingly.
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md]: The comment that replicas/resources/envs 'may vary accordingly' appears directly above the PyTorch-only `master` block, which is fixed at 1 replica. This ordering makes it easy to misread the note as applying to `master`. Moving the generic note into the framework-specific configuration section, or adding 'except for PyTorch `master`, which is always 1 replica', would remove the ambiguity.

File: keps/arena-v2/yaml-schema.md [hunk 18]
```diff
@@ -112,34 +134,34 @@ scheduling:                            # Optional
...
+      # topology_key: nvidia.com/gpu   # Only for target: pod
       match_expressions:
       - key: foo
         operator: In
         values:
         - bar
-      match_fields:
+      match_fields:                   # Only for target: node
       - key: foo
         operator: In
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:150]: `lifecycle.success_policy` is annotated 'TFJob only'. It would help to state whether the field is silently ignored for PyTorchJob/MPIJob or whether validation rejects it in Alpha. That saves users from guessing when they copy a TFJob example to another framework.

File: keps/arena-v2/yaml-schema.md [hunk 19]
```diff
@@ -367,39 +418,40 @@ PyTorchJob CRD
...
+| tensorflow | Worker | Yes | Chief/PS/Evaluator do not inherit worker's config |
+| mpi | Worker | Yes | Launcher does not use GPU, does not inherit worker's config |
+| deepspeed | Worker | Yes | Launcher does not use GPU, does not inherit worker's config |
+| horovod | Worker | Yes | Launcher does not use GPU, does not inherit worker's config |
 
-**MPI/DeepSpeed/Horovod Launchers do not inherit worker's GPU resources by default** — the Provider automatically sets independent resource config for the Launcher (typically no GPU) during CRD generation, avoiding GPU waste on Launcher pods.
+**MPI/DeepSpeed/Horovod Launchers do not inherit worker's config by default** — the Provider automatically sets independent resource config for the Launcher (typically no GPU) during CRD generation, avoiding GPU waste on Launcher pods.
 
-**PyTorch Master vs Worker**: Master uses the same `worker.resources` config (including GPU) as Worker. Master replicas is fixed at 1; Worker replicas = `worker.replicas - 1`.
+**PyTorch Master vs Worker**: Master runs as Worker. Master replica count is fixed at 1; **total GPU-using replicas** = `worker.replicas + 1`.
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md]: 'Master runs as Worker' is easy to parse as 'the master pod is a worker pod'. The intended meaning seems to be that the master uses the worker's image/resources by default. A clearer phrasing like 'Master inherits the worker image/resources by default' would avoid confusion.

**Comment by cheyang** [comment]
@gemini-code-assist /review

**Review by cheyang — APPROVED** [review]
/lgtm
/approve
